### PR TITLE
Yaml2json types

### PR DESCRIPTION
**Pull request by bitdivine**
# Motivation
`yaml2candid` is very incomplete and still in some places buggy.  Specifically, it does not support many Candid types and the opt test vectors (and implementation that was guided by those test vectors) are wrong.

# Changes
- Restructure the main `yaml2json` type match so that the compiler will complain when not all types are matched.
- Add YAML to Candid conversions for almost all missing Candid types.  Service and class are still unsupported.  I am running out of time and these are both complicated and not needed for my immediate goal.
- Add lots of tests.
- Fix the existing test vector that didn't include `opt` in the candid.

# Tests
- Fairly comprehensive unit tests are included for each individual type.
- Test coverage was checked with Tarpaulin; the tarpaulin report is generated by CI and can be checked.
- Most but not all error branches are tested.

# TODO in future
- Ensure that failures have useful error messages.
- Use proptests to ensure that no corner cases have been overlooked.
- Perhaps get more test vectors from real-life use to ensure that the test vectors are consistent with conventional usage.

### CODE REVIEW

File: src/yaml2candid/src/lib.rs [hunk 1]
```diff
@@ -67,8 +84,8 @@ impl Yaml2Candid {
     /// * YAML types do not match the IDL types.
     /// * `type_name` is not defined in the converter's `.did` file.
     pub fn convert(&self, typ: &IDLType, data: &YamlValue) -> anyhow::Result<IDLValue> {
```
Comment (bitdivine) [inline src/yaml2candid/src/lib.rs:86]: Here the match includes just the Candid type, so that the compiler will complain if there are any unsupported Candid types.

Below _many_ types that were previously unsupported have been added.

File: src/yaml2candid/src/lib.rs [hunk 2]
```diff
@@ -3,19 +3,36 @@
...
 use std::path::Path;
+#[cfg(test)]
+mod test;
 
 /// Converts YAML to Candid using a given did file.
 pub struct Yaml2Candid {
     /// The types that the converter supports, defined in an IDLProg interface definition.
     pub prog: IDLProg,
 }
+impl Default for Yaml2Candid {
```
Comment (bitdivine) [inline src/yaml2candid/src/lib.rs:25]: Simple default constructor needed for tests.  Included here as it may have more uses.

File: src/yaml2candid/Cargo.toml [hunk 3]
```diff
@@ -11,3 +11,7 @@ candid_parser = { workspace = true }
 serde = "1"
 serde_yaml = "0.9"
 anyhow = "1"
+num-bigint = { workspace = true }
+
+[dev-dependencies]
+pretty_assertions = "1.4.0"
```
Comment (bitdivine) [inline src/yaml2candid/Cargo.toml:17]: Provides a pretty diff in `assert_eq!()` errors.  This makes debugging much easier.

File: Cargo.toml [hunk 4]
```diff
@@ -16,3 +16,4 @@ version = "0.10.1"
 [workspace.dependencies]
 candid = { version = "0.10.8" }
 candid_parser = { version = "0.1.0" }
+num-bigint = { version = "0.4.5" }
```
Comment (bitdivine) [inline Cargo.toml:19]: Needed when converting BigNum types.

File: src/yaml2candid/tests/nns-sns-wasm/SnsInitPayload.idl [hunk 5]
```diff
@@ -1,60 +1,60 @@
 record {
```
Comment (bitdivine) [inline src/yaml2candid/tests/nns-sns-wasm/SnsInitPayload.idl:1]: In the .did file, many fields below are optional.  I believe that, strictly speaking, they should therefore be `opt` even if parsers might accept either.

**Review by chenyan-dfinity — APPROVED** [review]
